ibis/backends/flink/tests/test_ddl.py

```diff
@@ -534,3 +534,83 @@ def test_to_parquet(con, tmp_path, table_name):
     out_df = pd.read_parquet(out_path)
 
     tm.assert_frame_equal(source_df, out_df)
+
+
+def test_temporal_join(data_dir, con, tempdir_sink_configs):
+    schema = sch.Schema(
+        {
+            "id": dt.int32,
+            "bool_col": dt.bool,
+            "smallint_col": dt.int16,
+            "int_col": dt.int32,
+            "timestamp_col": dt.timestamp(scale=3),
+        }
+    )
+
+    path = f"{data_dir}/parquet/functional_alltypes.parquet"
+    df = pd.read_parquet(path)
+    df = df[list(schema.names)]
+    df = df.head()
+
+    # Create `table_left`
+    table_left_tempdir = tempfile.TemporaryDirectory()
+    table_left = con.create_table(
+        "table_left",
+        schema=schema,
+        tbl_properties=tempdir_sink_configs(table_left_tempdir.name),
+        watermark=ibis.watermark(
+            time_col="timestamp_col", allowed_delay=ibis.interval(seconds=15)
+        ),
+    )
+    con.insert(
+        "table_left",
+        obj=df,
+        schema=schema,
+    ).wait()
+
+    # Create `table_right`
+    table_right_tempdir = tempfile.TemporaryDirectory()
+    table_right = con.create_table(
+        "table_right",
+        schema=schema,
+        tbl_properties=tempdir_sink_configs(table_right_tempdir.name),
+        watermark=ibis.watermark(
+            time_col="timestamp_col", allowed_delay=ibis.interval(seconds=15)
+        ),
+        primary_key="id",
+    )
+    con.insert(
+        "table_right",
+        obj=df,
+        schema=schema,
+    ).wait()
+
+    # Join `table_left` and `table_right`
+    expr = table_left.temporal_join(
```

Comment:
How about a slightly different API `left.temporal_join(right.at_time(right.timestamp))` ?

That way the versioned table operation could hold the `at_time` property which could eliminate the majority of custom join logic.

Reply:
Thanks, I agree that this would be nicer. I added the API you suggested along side the initial one, just to keep the previous one as a reference until getting temporal join working.